marlowe/src/Language/Marlowe/Client/History.hs

```diff
@@ -0,0 +1,335 @@
+      addressFilter out =
+         do
+          guard
+            $ isNothing valAddress
+            || valAddress == Just (txOutAddress . tyTxOutTxOut $ tyTxOutRefOut out)
+          pure out
+      toMarlowe' (txOutRef, (citxOut, citx')) =
+        addressFilter
+          =<< toMarlowe citx' (toTxOut citxOut) txOutRef
+    mapMaybe toMarlowe'
+      <$> utxosTxOutTxFromTx citx
+
+
+-- | Extract the Marlowe state from a Marlowe-specific output.
+toMarloweState :: MarloweTxOutRef  -- ^ The Marlowe-specific output.
+               -> MarloweData      -- ^ The Marlowe data.
+toMarloweState = tyTxOutData . tyTxOutRefOut
+
+
+-- | Test whether a transaction created a Marlowe contract.
+creationTxOut :: MarloweParams          -- ^ The Marlowe validator parameters.
+              -> Address                -- ^ The Marlowe validator address.
+              -> ChainIndexTx           -- ^ The transaction to be checked.
+              -> Maybe MarloweTxOutRef  -- ^ The creation-transaction output and the contract, if any.
+creationTxOut MarloweParams{..} address citx =
+  do
+    -- Ensure that the transaction minted the role currency.
+    guard
+      . elem (ScriptHash $ unCurrencySymbol rolesCurrency) . M.keys
+      $ citx ^. citxScripts
+    -- Find the output to the script address, if any.
+    case filterOutputs address citx of
+      [creation] -> Just creation
+      _          -> Nothing
+
+
+-- | Find the Marlowe outputs at an address.
+filterOutputs :: Address            -- ^ The Marlowe address.
+              -> ChainIndexTx       -- ^ The transaction.
+              -> [MarloweTxOutRef]  -- ^ The outputs to the Marlowe address.
+filterOutputs address =
+  filter ((address ==) . txOutAddress . tyTxOutTxOut . tyTxOutRefOut)
+    . txMarloweData
+
+
+-- | Extract Marlowe data from a transaction.
+txMarloweData :: ChainIndexTx       -- ^ The transaction.
+              -> [MarloweTxOutRef]  -- ^ The outputs that have Marlowe data.
+txMarloweData citx =
+  catMaybes
+    $ uncurry (toMarlowe citx)
+    <$> txDatums citx
+
+
+-- | Make an output specific to Marlowe.
+toMarlowe :: ChainIndexTx           -- ^ The transaction.
+          -> TxOut                  -- ^ The output.
+          -> TxOutRef               -- ^ The output reference.
+          -> Maybe MarloweTxOutRef  -- ^ The Marlowe-specific output, if any.
+toMarlowe citx txOut txOutRef =
+  do
+    dh <- txOutDatumHash txOut
+    dat <- M.lookup dh $ citx ^. citxData
+    md <- fromBuiltinData $ getDatum dat
+    pure
+      . TypedScriptTxOutRef txOutRef
+      $ TypedScriptTxOut txOut md
+
+
+-- | Extract transaction outputs with datum.
+txDatums :: ChainIndexTx         -- ^ The transaction.
+         -> [(TxOut, TxOutRef)]  -- ^ The outputs that have datum.
+txDatums citx =
+  case citx ^. citxOutputs of
+    ValidTx txOuts -> filter (isJust . txOutDatumHash . fst)
+                      $ (fst &&& (TxOutRef (citx ^. citxTxId) . snd))
```

Comment:
It's a stylish remark, but could we try to use less funny Haskell and more boringly verbose Haskell? I mean, it's going to be 2 lines longer but easier to understand by non-veterans. And you could save of Control.Arrow import :)

Reply:
Yes, I'll replace the arrows with something easier to read.

Reply:
Fixed in cb1fba3cd5ae838b556925ab12230ed5b4b6bc35.